File: dart/bullet.dart
import 'image.dart';
import 'package:box2d/box2d_browser.dart';
import 'sprite.dart';
import 'dart:math';

class Bullet extends Sprite {
	var lastTime = 0;
	var bounceCooldown = 40;

	Bullet(Image img, World world, double startx, double starty, double startAngle, double startSpeed) {
		this.img = img;
		createBody(world, startx, starty);
		createShape();
		speed = -startSpeed;
		body.setTransform(new Vector2(startx, starty), radians(startAngle));

		body.bullet = true;
	}

	void createShape() {
		CircleShape dynamicCircle = new CircleShape();
		dynamicCircle.radius = 0.5;
		dimensions = new Vector2(10.0 * dynamicCircle.radius, 10.0 * dynamicCircle.radius);

		FixtureDef fixtureDef = new FixtureDef();
		fixtureDef.shape = dynamicCircle;
		fixtureDef.density = 1.0;
		fixtureDef.friction = 0.0;
		fixtureDef.userData = "Bullet";

		body.createFixture(fixtureDef);
	}

	void update() {
		updatePosition();
		checkCollision();
	}

	void checkCollision() {
		for (ContactEdge ce = body.contactList; ce != null; ce = ce.next) {
			if (ce.contact.touching) {
				var userData = ce.contact.fixtureA.userData;
				var otherUserData = ce.contact.fixtureB.userData;
				if (userData == "Tank" || otherUserData == "Tank") {
					destroy();
				} else if (userData == "Wall" || otherUserData == "Wall") {
					Body obstacle = ce.contact.fixtureA.body;
					if (otherUserData == "Wall") {
						obstacle = ce.contact.fixtureB.body;
					}

					double deltaX = body.position.x - obstacle.position.x;
					double deltaY = body.position.y - obstacle.position.y;

					double rightWallDis = deltaX - obstacle
				}
				break;
			}
		}
	}
}

File: js/bullet.js
Bullet = function(img, world, startx, starty, startAngle, startSpeed, id) {
	this.init(img, id);

	this.width = 0.05; this.height = 0.05;
	this.torque = 0;

	this.speed = startSpeed;
	this.createBody(world, startx, starty);
	this.createShape();

	this.body.SetAngle(startAngle);
	this.body.SetUserData("Bullet");
	this.body.SetBullet(true);

	this.bounces = 0;

	var vx = this.speed * Math.cos(this.body.GetAngle() - Math.PI / 2);
	var vy = this.speed * Math.sin(this.body.GetAngle() - Math.PI / 2);

	var velocity = new b2Vec2(vx, vy);
	this.body.ApplyImpulse(velocity, this.body.GetPosition());
}

Bullet.prototype = new Sprite();
Bullet.prototype.constructor = Bullet;

Bullet.prototype.update = function(dt) {
	Sprite.prototype.update.call(this, dt);
	this.checkCollision();
}

Bullet.prototype.createShape = function() {
	var fixDef = new b2FixtureDef;
	fixDef.density = 1.0;
	fixDef.friction = 0.0;
	fixDef.restitution = 1.0;

	fixDef.shape = new b2CircleShape;
	fixDef.shape.SetRadius(this.width);
	this.body.CreateFixture(fixDef);
}

Bullet.prototype.checkCollision = function() {
	for (var ce = this.body.GetContactList(); ce != null; ce = ce.next) {
		if (ce.contact.IsTouching()) {
			var userData = ce.contact.GetFixtureA().GetBody().GetUserData();
			var otherUserData = ce.contact.GetFixtureB().GetBody().GetUserData();

			if (userData === "Tank" || otherUserData === "Tank") {
				if (typeof io.sockets !== 'undefined') {
					this.destroy();
				} else {
					addExplosion(this.body.GetPosition().x * scale, this.body.GetPosition().y * scale);
				}
			} else {
				this.bounces++;

				if (this.bounces > 4) {
					if (typeof io.sockets !== 'undefined') {
						this.destroy();
					}
				}
			}
		}
	}
}

if (typeof module !== 'undefined') module.exports = Bullet;
